File: 508-most-frequent-subtree-sum/most-frequent-subtree-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    unordered_map<int,int> map;
    int maxFrequency=0;
    int solve(TreeNode* root)
    {
        if(!root)
        return 0;

        int sum = root->val;
        sum += solve(root->left);
        sum += solve(root->right);

        map[sum]++;
        maxFrequency=max(maxFrequency,map[sum]);
        
        return sum;
    }

    vector<int> findFrequentTreeSum(TreeNode* root) 
    {
        vector<int> ans;
        solve(root);

        for(auto x : map)
        {
            if(x.second==maxFrequency)
            ans.push_back(x.first);
        }
        return ans;
    }
};

File: 508-most-frequent-subtree-sum/most-frequent-subtree-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
private Map<Integer, Integer> subtreeSumMap;
    private int maxFrequency;
    private int subtreeSum(TreeNode root) {
        if(root == null) {
            return 0;
        }

        int sum = root.val + subtreeSum(root.left) + subtreeSum(root.right);
        
        subtreeSumMap.put(sum, subtreeSumMap.getOrDefault(sum, 0) + 1);
        maxFrequency = Math.max(maxFrequency, subtreeSumMap.get(sum));
        
        return sum;
    }

    public int[] findFrequentTreeSum(TreeNode root) {
        maxFrequency = 0;
        subtreeSumMap = new HashMap<>();
        subtreeSum(root);
        
        List<Integer> sumList = new ArrayList<>();
        for(Map.Entry<Integer, Integer> entry : subtreeSumMap.entrySet()) {
            if(maxFrequency == entry.getValue()) {
                sumList.add(entry.getKey());
            }
        }
        return sumList.stream().mapToInt(Integer::intValue).toArray();
    }
}